doc/arch/adr-010-learn-to-rank.md

```diff
@@ -0,0 +1,297 @@
+following reasons:
+
+- We think that it will improve search relevancy for citizens, enabling
+  them to find what they need on GOV.UK
+- We have validated our assumptions with a spike
+- LTR is a well trodden path, and other organisations have written about
+  their implementations with varying levels of openness (e.g. Wikimedia)
+- GOV.UK has implemented other machine learning projects with success
+  (related links, topic taxonomy), and we have a data science team which
+  the search team has already worked with
+- TensorFlow provides an off-the-shelf ranking module, which solves
+  some of the machine learning problems of LTR for us
+- Manual relevancy tuning is not scalable, and won't get us to the same
+  level of relevancy as LTR
+
+## Implementation plan
+
+We have decided to implement Learning To Rank by following on from
+the work started in the [validation spike][].
+
+As we are at the beginning of the project, this section may not
+accurately reflect the final implementation.
+
+[validation spike]: https://github.com/alphagov/search-api/pull/1768
+
+### Machine learning library
+
+We have decided to use [TensorFlow's ranking library][] for our
+implementation of Learning To Rank.
+
+TensorFlow is well supported, and the ranking library is under active development.
+
+We considered using the [elasticsearch-learning-to-rank][] plugin.
+However, AWS Elasticsearch does not permit us to install custom plugins.
+
+We also considered PyTorch. However, this would require writing much more
+custom machine learning code, as PyTorch don't currently have a ranking library.
+
+[TensorFlow's ranking library]: https://github.com/tensorflow/ranking
+[elasticsearch-learning-to-rank]: https://github.com/o19s/elasticsearch-learning-to-rank
+
+### Language
+
+We would like to keep the code for this project in Ruby as much as
+possible. Search API is written in Ruby and it is the primary language
+used at GOV.UK. Necessarily some machine learning code will need to be
+written in Python because TensorFlow is a Python library.
+
+### Dockerised model
+
+We will use TensorFlow Serving to make a re-ranker model available
+to Search API on search machines.
+
+We will run a docker container (containing our pre-trained model) on
+search machines to which Search API can send requests.
+
+Should the docker container fail, Search API should be resilient to this,
+and fall back to the default ranking algorithm, with results not
+re-ranked by the model.
+
+### Training
+
+**Training the model** should happen on search machines on a schedule.
+An instance of Search API will train a model using the latest relevancy
+judgements stored in S3, and will then upload a new version of the model to S3.
+On a schedule all search machines will pull the latest version of the model
+and push the new version to the locally running TF Serving container.
+
+**Training data** is made up of relevancy judgements and document features.
```

Comment:
Is it useful to perform some spot-check validation on a sample of relevancy judgements over time? It's not practical to inspect everything and we assume that the results users are clicking on are the best results for the query they've entered, but if we check a selection of these from every so often we can help to validate that we're training the model appropriately. 

Touching on what @sihugh mentioned earlier on also about gaming the system, do we have any thoughts about how we can ensure the model isn't misled by relevancy judgements through the actions of a malicious user?

Reply:
💯 Perhaps some of this information can be surfaced in the publishing tools as well? If publishers can see the judgement lists for specific queries, maybe they can flag where content is completely missing.

Reply:
That's definitely worth investigating! I've added a task to our backlog to find a way for editors to spot check the autogenerated relevancy judgements.

> do we have any thoughts about how we can ensure the model isn't misled by relevancy judgements through the actions of a malicious user?

I added some thoughts on gaming this in a comment above. I think it's worth investigating - there's a card on our backlog for this too.